src/tamp_improv/approaches/improvisational/policies/multi_rl.py

```diff
@@ -127,9 +106,70 @@ def train(self, env: gym.Env, train_data: TrainingData | None) -> None:
 
         print(f"\nTrained {len(self._policies)} specialized policies")
 
+    def _get_policy_key(self, context: PolicyContext) -> str:
+        """Create a unique key for a policy based on the context."""
+        # Get ground atoms as strings to preserve object information
+        source_atoms_str = sorted([str(atom) for atom in context.current_atoms])
+        target_atoms_str = sorted([str(atom) for atom in context.preimage])
+
+        # Create hash of the source and target atoms
+        source_hash = hashlib.md5("|".join(source_atoms_str).encode()).hexdigest()[:8]
+        target_hash = hashlib.md5("|".join(target_atoms_str).encode()).hexdigest()[:8]
+
+        # Include source and target node IDs if available
+        source_id = context.info.get("source_node_id", "")
+        target_id = context.info.get("target_node_id", "")
+
+        if source_id != "" and target_id != "":
+            return f"n{source_id}-to-n{target_id}_{source_hash}_{target_hash}"
+        return f"{source_hash}_{target_hash}"
+
+    def _find_matching_policy(self, context: PolicyContext) -> str | None:
+        """Find a matching policy based on relevant atoms of the shortcut."""
+        source_atoms = context.current_atoms
+        target_atoms = context.preimage
+        added_atoms = target_atoms - source_atoms
+        deleted_atoms = source_atoms - target_atoms
+
+        # Try exact key match first
+        key = self._get_policy_key(context)
+        if key in self._policies:
+            return key
+
+        # Try structural matching
+        for policy_key, pattern_info in self._policy_patterns.items():
+            train_added = pattern_info["added_atoms"]
+            train_deleted = pattern_info["deleted_atoms"]
+
+            # Check predicate subsets
+            if not self._check_predicate_subset(added_atoms, train_added):
+                continue
+            if not self._check_predicate_subset(deleted_atoms, train_deleted):
+                continue
+
+            # Pool atoms together for substitution finding
```

Comment:
hmm this seems a bit risky... what I'm about to propose is ugly coding-wise, but I think it would be safer to transform these predicates so that an add effect of `Holding(block1)` becomes `ADD-Holding(block1)` whereas a delete effect becomes `DEL-Holding(block1)`.